Make it easier to test csrf & security component protected actions.

markstory opened the pull request:
Make it much simpler for developers to use IntegrationTestCase with controllers protected by SecurityComponent and CsrfComponent.

I've made the `Token` data generation optional and require an explicit method call, as it adds some overhead, and the presence of that data could be unexpected by developers.

Refs #7004

markstory committed:
Automatically append CSRF tokens to requests.

Add the cookie and POST data to requests if they are not already
defined. This lets developers focus on application code and not worry
about passing CSRF token validation.

Refs #7004

markstory committed:
Split security token generation out of FormHelper.

Having a separate trait will allow reuse in IntegrationTestCase, which
will let us making testing controllers that use SecurityComponent
simple.

Refs #7004

markstory committed:
Remove useless property.

The modelClass property is not used anymore.

markstory committed:
Add easy way to test actions using security component.

By calling enableSecurityToken() developers can opt-into having
automatic security tokens generated for their requests. This makes
integration testing with SecurityComponent more easy as developers don't
have to calculate these tokens themselves.

Refs #7004

steefaan commented on src/TestSuite/IntegrationTestCase.php:
```diff
@@ -343,11 +367,11 @@ protected function _buildRequest($url, $method, $data)
         $props = [
             'url' => $url,
-            'post' => $data,
+            'post' => $this->_addTokens($url, $method, $data),
```
This will add the security token conditionally but adds the `_csrfToken` in any case or I overlooked something? What happens if the user doesn't use `CsrfComponent`? The `_csrfToken` itself doesn't hurt but it will change the request data.

markstory replied:
Do you think there should be an `enableCsrfToken()` method as well?

steefaan replied:
Yes because it's more consistent in my opinion. Let the user decide what kind of data he wants to inject to the request instead of force the `_csrfToken`. The forced `_csrfToken` may ends up with some failing test cases in existing applications and would be a BC break if you check against the number of setted cookies or similar things for whatever reason.

markstory replied:
Sounds like a good idea, I came to the same conclusion last night. The consistency is more important than the automatic behavior.

markstory committed:
Restore removed property.

This property stops missing table errors.

markstory committed:
Make CSRF token assignment use a method as well.

This makes SecurityComponent and CsrfComponent work the same in the
IntegrationTestCase.

Refs #7004

steefaan commented:
:+1:
